File: controllers/vendor.controller.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const Vendor = require('../models/Vendor');
const MenuItem = require('../models/MenuItem');
const Order = require('../models/Order');

const signToken = (vendor) =>
  jwt.sign(
    { id: vendor._id, role: 'vendor' },
    process.env.JWT_SECRET || 'food_genie_secret',
    { expiresIn: '7d' }
  );

// GET /api/vendors
exports.getAllVendors = async (req, res) => {
  try {
    const vendors = await Vendor.find({ isActive: true, isApproved: true });
    res.json(vendors);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// GET /api/vendors/:id
exports.getVendorById = async (req, res) => {
  try {
    const vendor = await Vendor.findById(req.params.id);
    if (!vendor) return res.status(404).json({ message: 'Vendor not found.' });
    res.json(vendor);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// POST /api/vendors/register
exports.registerVendor = async (req, res) => {
  try {
    const { name, email, password, phone, address, cuisine } = req.body;
    const exists = await Vendor.findOne({ email });
    if (exists) return res.status(400).json({ message: 'Email already in use.' });

    const hashed = await bcrypt.hash(password, 12);
    const vendor = await Vendor.create({
      name,
      email,
      password: hashed,
      phone,
      address,
      cuisine,
      isApproved: false // Requires admin approval
    });

    const vendorResponse = vendor.toObject();
    delete vendorResponse.password;

    res.status(201).json({
      token: signToken(vendor),
      vendor: vendorResponse
    });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// PUT /api/vendors/:id
exports.updateVendor = async (req, res) => {
  try {
    if (req.user.role !== 'admin' && req.user.id !== req.params.id) {
      return res.status(403).json({ message: 'Access denied.' });
    }

    const vendor = await Vendor.findByIdAndUpdate(req.params.id, req.body, { new: true });
    if (!vendor) return res.status(404).json({ message: 'Vendor not found.' });
    res.json(vendor);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// GET /api/vendors/:id/menu
exports.getVendorMenu = async (req, res) => {
  try {
    const menu = await MenuItem.find({ vendor: req.params.id });
    res.json(menu);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// GET /api/vendors/:id/orders
exports.getVendorOrders = async (req, res) => {
  try {
    if (req.user.role !== 'admin' && req.user.id !== req.params.id) {
      return res.status(403).json({ message: 'Access denied.' });
    }

    const orders = await Order.find({ vendor: req.params.id })
      .populate('customer', 'name email phone')
      .populate('rider', 'name phone')
      .sort('-createdAt');
    res.json(orders);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// PATCH /api/vendors/:id/status
// Body contains { status: 'open' | 'closed' }
exports.updateVendorStatus = async (req, res) => {
  try {
    if (req.user.role !== 'admin' && req.user.id !== req.params.id) {
      return res.status(403).json({ message: 'Access denied.' });
    }

    const { status } = req.body;
    if (!['open', 'closed'].includes(status)) {
      return res.status(400).json({ message: 'Invalid status. Must be open or closed.' });
    }

    const vendor = await Vendor.findByIdAndUpdate(
      req.params.id,
      { status },
      { new: true }
    );
    if (!vendor) return res.status(404).json({ message: 'Vendor not found.' });
    res.json(vendor);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

File: controllers/vendor_controller.py
from fastapi import HTTPException, status
from sqlalchemy import select, update
from database import AsyncSessionLocal, serialize_doc
from middleware.auth import hash_password, create_access_token
from models.vendor_model import VendorRegister, VendorUpdate, VendorStatusUpdate
from orm_models import Vendor, User, Food

async def get_all_vendors():
    try:
        async with AsyncSessionLocal() as session:
            res = await session.execute(select(Vendor).where(Vendor.is_active == True))
            vendors = res.scalars().all()
            serialized = serialize_doc(list(vendors))
            for v in serialized:
                v.pop("password", None)
            return serialized
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch vendors: {str(e)}"
        )

async def get_vendor_by_id(vendor_id: str):
    async with AsyncSessionLocal() as session:
        res = await session.execute(select(Vendor).where(Vendor.id == str(vendor_id)))
        vendor = res.scalar_one_or_none()
        if not vendor:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vendor not found")

        serialized = serialize_doc(vendor)
        serialized.pop("password", None)
        return serialized

async def register_vendor(data: VendorRegister):
    try:
        email_clean = data.email.lower()
        async with AsyncSessionLocal() as session:
            res_v = await session.execute(select(Vendor).where(Vendor.email == email_clean))
            res_u = await session.execute(select(User).where(User.email == email_clean))
            if res_v.scalar_one_or_none() or res_u.scalar_one_or_none():
                raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Email is already in use")

            hashed_pw = hash_password(data.password)
            vendor_obj = Vendor(
                name=data.name,
                email=email_clean,
                password=hashed_pw,
                city=data.city or "Vehari",
                phone=data.phone,
                category=data.category or "Pakistani",
                cuisine=data.cuisine or "General",
                address=data.address or "Vehari, Pakistan",
                status="open",
                rating=4.5,
                is_active=True,
                is_approved=True,
                logo="",
                cover_image=""
            )

            session.add(vendor_obj)
            await session.commit()
            await session.refresh(vendor_obj)

            serialized = serialize_doc(vendor_obj)
            serialized.pop("password", None)

            token = create_access_token({"id": serialized["id"], "role": "vendor"})

            return {
                "token": token,
                "role": "vendor",
                "vendor": serialized
            }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Vendor registration failed: {str(e)}"
        )

async def update_vendor(vendor_id: str, data: VendorUpdate, current_user: dict):
    if current_user.get("role") != "admin" and current_user.get("id") != str(vendor_id):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Unauthorized to update vendor profile")

    async with AsyncSessionLocal() as session:
        res = await session.execute(select(Vendor).where(Vendor.id == str(vendor_id)))
        vendor = res.scalar_one_or_none()
        if not vendor:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vendor not found")

        update_dict = {k: v for k, v in data.dict(exclude_unset=True).items() if v is not None}
        for k, v in update_dict.items():
            if hasattr(vendor, k):
                setattr(vendor, k, v)

        await session.commit()
        await session.refresh(vendor)

        serialized = serialize_doc(vendor)
        serialized.pop("password", None)
        return serialized

async def get_vendor_menu(vendor_id: str):
    try:
        async with AsyncSessionLocal() as session:
            res = await session.execute(select(Food).where(Food.vendor_id == str(vendor_id)))
            foods = res.scalars().all()
            return serialize_doc(list(foods))
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to fetch vendor menu: {str(e)}"
        )

async def update_vendor_status(vendor_id: str, data: VendorStatusUpdate, current_user: dict):
    if current_user.get("role") != "admin" and current_user.get("id") != str(vendor_id):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Unauthorized to update vendor status")

    status_val = data.status.value if hasattr(data.status, "value") else str(data.status)
    async with AsyncSessionLocal() as session:
        res = await session.execute(select(Vendor).where(Vendor.id == str(vendor_id)))
        vendor = res.scalar_one_or_none()
        if not vendor:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Vendor not found")

        vendor.status = status_val
        await session.commit()
        await session.refresh(vendor)

        serialized = serialize_doc(vendor)
        serialized.pop("password", None)
        return serialized
